modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,371 @@
+package usage
+
+import (
+	"math"
+	"math/bits"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/prometheus/util/strutil"
+
+	v1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+)
+
+const (
+	tenantLabel  = "tenant"
+	trackerLabel = "tracker"
+)
+
+var emptyHash = hash(nil, nil)
+
+type tenantLabelsFunc func(string) []string
+
+type bucket struct {
+	// Configuration
+	descr  *prometheus.Desc // Configuration can change over time so it is captured with the bucket.
+	labels []string
+
+	// Runtime data
+	bytes uint64
+	unix  int64
+}
+
+func (b *bucket) Inc(bytes uint64, unix int64) {
+	b.bytes += bytes
+	b.unix = unix
+}
+
+type tenantUsage struct {
+	series      map[uint64]*bucket
+	constLabels prometheus.Labels
+
+	// Buffers for Observe
+	dimensions []string // Originally configured dimensions
+	labels     []string // Sanitized dimensions => final labels
+	values     []string // Used to capture values and guaranteed to be a matching length.
+}
+
+// GetBuffersForDimensions takes advantage of the fact that the configuration for a tracker
+// changes slowly.  Reuses buffers from the previous call when the dimensions are the same.
+func (t *tenantUsage) GetBuffersForDimensions(dimensions []string) ([]string, []string) {
+	reset := false
+
+	if len(dimensions) == len(t.dimensions) {
+		// Any change to dimensions?
+		for i := range dimensions {
+			if dimensions[i] != t.dimensions[i] {
+				reset = true
+				break
+			}
+		}
+	} else {
+		reset = true
+	}
+
+	if reset {
+		t.dimensions = dimensions
+		t.labels = make([]string, len(dimensions))
+		t.values = make([]string, len(dimensions))
+		for i, d := range dimensions {
+			t.labels[i] = strutil.SanitizeFullLabelName(d)
+		}
+	}
+
+	return t.labels, t.values
+}
+
+func (t *tenantUsage) getSeries(labels, values []string, maxCardinality int) *bucket {
+	h := hash(labels, values)
+
+	b := t.series[h]
+	if b == nil {
+		// Before creating a new series, check for cardinality limit.
+		if len(t.series) >= maxCardinality {
+			// Overflow
+			// It goes into the unlabeled bucket
+			// TODO - Do we want to do something else?
```

Comment:
i really like this solution and can't come up with a better one off the top of my head. you can clearly see the severity of the unknown data and choose whether or not to pick a different label set.

Reply:
This behavior was changed slightly.  Instead of going to an unlabeled bucket, it substitutes `__overflow__` for every configured dimension. So if your dimensions are "foo" and "bar", the overflow series will be `{foo="__overflow__", bar="__overflow__"}`.  This makes it more explicit, and simplifies the downstream querying and reporting (and the same approach is being taken in the other databases).